include/zephyr/async.h

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright (c) 2022 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ *
+ * @brief Public APIs for Async utilities
+ *
+ * Functions and datatypes useful for safely communicating to a thread or work
+ * queue of an asynchronous functions completion.
+ */
+
+#ifndef ZEPHYR_INCLUDE_ASYNC_H_
+#define ZEPHYR_INCLUDE_ASYNC_H_
+
+#include <zephyr/kernel.h>
+
+/**
+ * @brief Async operation callback type
+ *
+ * Common function pointer type used for asynchronous callbacks
+ * and implemented for some commonly used IPC mechanisms.
+ *
+ * May also be implemented by an application with the understanding
+ * that the callback is most likely to be done in the ISR context
+ * with the limitations and restrictions that imposes.
+ *
+ * @param result The result of the asynchronous call
+ * @param userdata Generic userdata requested to be tied
+ *        to the callback.
+ */
+typedef void (*async_callback_t)(int result, void *userdata);
+
+#ifdef CONFIG_ASYNC
+
+/**
+ * Semaphore completion object with return code.
+ */
+struct async_sem {
+	/** Semaphore that will be given on completion */
+	struct k_sem *sem;
+
+	/** Result of async call */
+	int result;
+};
+
+/**
+ * Asynchronous semaphore completion callback
+ *
+ * @param result Result to assign to the async_sem
+ * @param userdata Pointer to a struct async_sem
+ */
+void async_semaphore_fn(int result, void *userdata);
+
+/**
+ * Signal completion callback
+ *
+ * @param result Result to assign to the signal
+ * @param userdata Pointer to a struct k_poll_signal
+ */
+void async_signal_fn(int result, void *userdata);
+
+/**
+ * Mutex completion object with return code.
+ */
+struct async_mutex {
+	/** Mutex that will be unlocked on  completion */
+	struct k_mutex *mutex;
+
+	/** Result of async call */
+	int result;
+};
+
+/**
+ * Asynchronous mutex completion callback
+ *
+ * @param result Result to assign to async_mutex
+ * @param userdata Pointer to struct async_mutex
+ */
+void async_mutex_fn(int result, void *userdata);
+
+/**
+ * Work completion object with return code.
+ */
+struct async_work {
+	/** Work that will be submitted on completion */
+	struct k_work work;
```

Comment:
why not a pointer like the other?

Reply:
For k_work to get the result, I used CONTAINER_OF. Probably could all be non-pointer with a macro for each to make it easy to define statically each one.

Reply:
the user knows that the work is tied to this async struct (as for the other types, specifically the sem and mutex to get the result) so I don't see any real gain in user CONTAINER_OF here. 

I'd rather generalize on one struct made of pointers instead of specializing for each notifier type. Unless there is really a drawback in this approach I am missing